feat: added parser for model definitions

HassanAkbar opened the pull request:
Added parser for model definition

for -> https://github.com/lutaml/lutaml-model/issues/285

HassanAkbar committed:
feat: added parser for model definitions

HassanAkbar committed:
handle requires

ronaldtse commented:
Thanks @HassanAkbar , we will need to provide documentation and also replace (and complete) the old Lutaml::Uml parsing implementation.

HassanAkbar commented:
> Thanks @HassanAkbar , we will need to provide documentation and also replace (and complete) the old Lutaml::Uml parsing implementation.

Yes, I will do that once this functionality is complete, the previous implementation is more ruby based so I will have to convert it into transformer class which is recommended by parslet.

ronaldtse commented:
@HassanAkbar I’m not a fan of the Parslet transformer architecture: there is no need to intentionally adopt it. My experience is it makes things hard to debug and its signature based rules make it hard to handle a number of cases.

HassanAkbar commented:
> @HassanAkbar I’m not a fan of the Parslet transformer architecture: there is no need to intentionally adopt it. My experience is it makes things hard to debug and its signature based rules make it hard to handle a number of cases.

Then I think these can be merged into the UML rules, I'll update this PR by merging it with current UML implementation.

HassanAkbar committed:
some more changes

HassanAkbar committed:
Updated parslet rules and classes

HassanAkbar committed:
updated parsing rules and conversion methods

HassanAkbar committed:
Merges lml parser to uml

HassanAkbar committed:
Merges classes for lml and uml parsers

HassanAkbar committed:
Added specs and test files for lml parsing

HassanAkbar committed:
Removes old lml parser

HassanAkbar committed:
Adds support for instances, collections, imports, exports and instance inheritance.

HassanAkbar committed:
Fixes the parser for enum definitions

HassanAkbar committed:
Refactors code and improves specs

HassanAkbar committed:
Adds files for lml seperately

HassanAkbar committed:
Merge branch 'main' of github.com:lutaml/lutaml into adding_lml_parser_for_lutaml_model_definitions

HassanAkbar committed:
Removes changes of lml parser from dsl

HassanAkbar committed:
Adds lml parser classes

HassanAkbar committed:
Removes pry

HassanAkbar committed:
Adds specs for lml parser

HassanAkbar committed:
removes pry from spec

HassanAkbar committed:
Inherits uml datatype

HassanAkbar committed:
adds extended to attribute

HassanAkbar committed:
Refactors lml classes and adds data preprocessing

HassanAkbar committed:
removes pry

HassanAkbar committed:
Fixes inner attributes and instances issue and refactors code

HassanAkbar commented:
@ronaldtse I have updated the PR to separate the `LML` from `UML` by extending the UML classes. Can you take a look and let me know if this is how we want to do it? or do we need to change something here? Thanks

HassanAkbar committed:
Merge origin/main: resolve conflicts with new parser architecture

- Gemfile, lutaml.gemspec: take main's versions (branch deltas obsolete)
- lib/lutaml/parser.rb: delete per main's god-method elimination (3517fdc);
  LML callers use Lutaml::Lml::Parser.parse directly
- lib/lutaml/uml/parsers/dsl.rb: keep branch grammar extensions +
  create_document hook; adopt main's ParsingError < Lutaml::Error and
  require cleanup; attribute_type keeps optional colon
- lib/lutaml.rb: register Lml in the autoload registry (replaces the
  loading previously done via deleted lib/lutaml/parser.rb)

HassanAkbar committed:
fix: correct LML model type declarations and parser iteration bugs

HassanAkbar committed:
resolve copilot comments and fix linter issues

HassanAkbar commented on lib/lutaml/uml/parsers/dsl.rb:
```diff
@@ -50,37 +56,80 @@ def parse(input_file, _options = {})
+
+        rule(:quotes) { match['"\''] }
+        rule(:quotes?) { quotes.maybe }
+        rule(:space) { match("\s") }
```
Not a bug: in a double-quoted Ruby string "\s" is the space character (0x20), not "s"

HassanAkbar commented on lib/lutaml/lml/data_processor.rb:
```diff
@@ -0,0 +1,186 @@
+module Lutaml
```
Applied

HassanAkbar commented on lib/lutaml/uml/parsers/dsl.rb:
```diff
@@ -124,51 +214,106 @@ def parse(input_file, _options = {})
             .as(:attributes)
         end
 
+        rule(:keyword_type_argument) do
+          str("type") >>
+            spaces? >>
+            match["[^\s\n\r]"].repeat(1).as(:type) >>
+            whitespace?
+        end
```
False premise: "\s" in a double-quoted Ruby string is a space character, so this pattern compiles to `[^ \n\r]`

HassanAkbar commented on lib/lutaml/uml/parsers/dsl.rb:
```diff
@@ -124,51 +214,106 @@ def parse(input_file, _options = {})
+            whitespace?
+        end
+
+        rule(:keyword_any_argument) do
+          spaces? >>
+            match("[^\s\n\r]").repeat(1).as(:name) >>
+            spaces >>
+            str("=").maybe >>
+            spaces? >>
+            attribute_value.as(:value) >>
+            whitespace?
```
Same as above

HassanAkbar commented on lib/lutaml/uml/parsers/dsl.rb:
```diff
@@ -405,8 +576,109 @@ def parse(input_file, _options = {})
+            str("}") >> whitespace?
+        end
+
+        rule(:import_definition) do
+          match("[^\s\n\r]").repeat(1).as(:format_type) >> spaces? >> quoted_string.as(:file) >> whitespace? >>
+            str("{") >> whitespace? >>
+            attributes >> whitespace? >>
+            str("}") >> whitespace?
```
Same false premise: "\s" evaluates to a space character in double quotes, so `format_type` already matched `[^ \n\r]`

HassanAkbar committed:
fix: address Copilot review

HassanAkbar commented on lib/lutaml/lml/parser.rb:
```diff
@@ -0,0 +1,247 @@
+        ::Lutaml::Lml::Diagram.new.tap do |model|
+          set_lml_model(model, hash)
+        end
+      end
```
Resolved by deletion rather than adding a model

HassanAkbar commented on lib/lutaml/uml/parsers/dsl.rb:
```diff
@@ -50,37 +56,80 @@ def parse(input_file, _options = {})
-        rule(:name) { match["a-zA-Z0-9 _-"].repeat(1) }
-        rule(:newline) { str("\n") >> str("\r").maybe }
+        rule(:newline) { match('[\r\n]') }
+
+        rule(:quoted_string) do
+          str('"') >> (str('"').absent? >> any).repeat.as(:string) >> str('"')
+        end
```
Fixed.

HassanAkbar commented on lib/lutaml/lml/top_element_attribute.rb:
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+require "lutaml/uml/class"
+require "lutaml/lml/cardinality"
```
Fixed.

HassanAkbar commented on lib/lutaml/lml/top_element_attribute.rb:
```diff
@@ -0,0 +1,18 @@
+      attribute :value, Lutaml::Model::Type::Value
+      attribute :attributes, TopElementAttribute, collection: true, default: []
+      attribute :extended, :boolean
+      attribute :instances, Instance, collection: true, default: []
```
Applied

HassanAkbar commented on lib/lutaml/lml/instance_collection.rb:
```diff
@@ -0,0 +1,17 @@
+      attribute :instances, Instance, collection: true, default: []
+      attribute :imports, InstancesImport, collection: true, default: []
+      attribute :exports, InstancesExport, collection: true, default: []
+      attribute :collections, Collection, collection: true, default: []
```
Applied

HassanAkbar commented on lib/lutaml/lml/instances_export.rb:
```diff
@@ -0,0 +1,12 @@
+  module Lml
+    class InstancesExport < Lutaml::Model::Serializable
+      attribute :format_type, :string
+      attribute :attributes, TopElementAttribute, collection: true, default: []
```
Applied

HassanAkbar commented on lib/lutaml/lml/instances_import.rb:
```diff
@@ -0,0 +1,13 @@
+    class InstancesImport < Lutaml::Model::Serializable
+      attribute :format_type, :string
+      attribute :file, :string
+      attribute :attributes, TopElementAttribute, collection: true, default: []
```
Applied

HassanAkbar committed:
fix: empty instance bodies produced a phantom attribute instead of an empty list

Copilot flagged nil entries from non-Hash tokens in
process_attributes_array. The stated mechanism is wrong — parslet drops
unnamed whitespace matches when mixed with named captures, so populated
:attributes arrays only ever contain hashes (verified on the raw parse
tree). But the adjacent edge case is real: for an EMPTY body the
capture is [], the all? branch is vacuously true, and the method
returned {} — which lutaml-model coerced into a phantom
TopElementAttribute, breaking .find/.map on instance.attributes.

Guard: empty input returns []. Spec covers empty and whitespace-only
instance bodies.

HassanAkbar commented on lib/lutaml/lml/data_processor.rb:
```diff
@@ -0,0 +1,188 @@
+            end
+          end
+          hash
+        else
+          obj.map { |item| process_attributes(item) }
+        end
```
**stated mechanism is wrong**. I inspected the raw parse tree with blank lines, stray whitespace, and comments between attributes: parslet drops unnamed whitespace matches when they're mixed with named captures, so populated `:attributes` arrays **only ever contain Hashes**.

HassanAkbar commented:
This has been moved to https://github.com/lutaml/lutaml-lml. This one is not needed now.
